File: 2025/3주차/BOJ28278_스택_2.cpp
#include <iostream>
using namespace std;

struct Node {
    int data;
    Node* next;
    Node(int x) : data(x), next(nullptr) {}
};

class Stack {
private:
    Node* topNode;
    int count;

public:
    Stack() : topNode(nullptr), count(0) {}

    ~Stack() {
        while(topNode != nullptr){
            Node* temp = topNode;
            topNode = topNode->next;
            delete temp; 
        }
    }


    /*
    설명: 새 노드를 생성하고,
    push : 새 노드의 next가 기존의 topNode(맨 위)를 가리키게 한 뒤,
    topNode를 새 노드로 바꿉니다.
    count를 1 증가시킵니다.
    정상 동작: 스택의 LIFO(후입선출) 구조를 잘 구현하고 있습니다.
    */

    void push(int x) {
        Node* newNode = new Node(x);
        newNode->next = topNode; // 항상 맨 앞에 추가
        topNode = newNode;
        count++;
    }

    int pop() {
        if (topNode == nullptr) return -1;
        int ret = topNode->data;
        Node* temp = topNode;
        topNode = topNode->next;
        delete temp;
        count--;
        return ret;
    }

    int size() { return count; }

    int empty() { return count == 0 ? 1 : 0; }

    int top() {
        if (topNode  == nullptr) return -1;
        return topNode->data;
    }
};


int main(){
    ios_base::sync_with_stdio(0);
    cin.tie(0); cout.tie(0);

    Stack* stack = new Stack();

    int N;
    cin >> N;

    while(N--){
        int cmd;
        cin >> cmd;
        switch(cmd) {
            case 1: {
                int x; cin >> x;
                stack->push(x);
                break;
            }
            case 2: {
                if (stack->empty()){ 
                    cout << -1 << "\n"; break; }
                int res = stack->top();
                stack->pop();
                cout << res << "\n";
                break;
            }
            case 3: {
                cout << stack->size() << "\n";
                break;
            }
            case 4: {
                cout << ((stack->empty())? 1 : 0) << "\n"; break;
            }
            case 5: {
                if(stack->empty()){ cout << -1 << "\n"; break;}
                cout << stack->top() << "\n";
                break;
            }
        }

    }
}

File: 2025/3주차/BOJ28278_스택_2.py
'''🧩 전체 설계 흐름 요약
연결 리스트로 스택의 구조 구현

객체지향적으로 동작을 Stack 클래스에 분리

문제 입출력 로직은 main()에서 관리

직접 실행과 import 구분을 위해 __name__ == "__main__" 사용'''

import sys
input = sys.stdin.readline

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None  # 다음 노드를 가리킴

class Stack:
    def __init__(self):
        self.top = None  
        self.count = 0

    def push(self, x):
        new_node = Node(x)
        new_node.next = self.top  # 새 노드가 현재 top을 가리킴
        self.top = new_node       # top을 새 노드로 갱신
        self.count += 1

    def pop(self):
        if self.empty():
            return -1
        popped_data = self.top.data
        self.top = self.top.next  # top을 다음 노드로 이동
        self.count -= 1
        return popped_data

    def size(self):
        return self.count

    def empty(self):
        return 1 if self.top is None else 0

    def top_value(self):
        if self.empty():
            return -1
        return self.top.data


'''
| 함수            | 설명                           |
| ------------- | ---------------------------- |
| `push(x)`     | 새 `Node(x)` 생성 후 top 앞에 붙임   |
| `pop()`       | `top` 노드를 제거하고 다음 노드로 top 이동 |
| `empty()`     | `top`이 `None`인지 확인           |
| `top_value()` | 현재 top의 data를 반환 (없으면 -1)    |
'''
def main():
    n = int(input())
    s = Stack()

    for _ in range(n):
        cmd = input().split()
        if cmd[0] == '1':
            s.push(int(cmd[1]))
        elif cmd[0] == '2':
            print(s.pop())
        elif cmd[0] == '3':
            print(s.size())
        elif cmd[0] == '4':
            print(s.empty())
        elif cmd[0] == '5':
            print(s.top_value())

'''이 파일이 직접 실행될 때만 main() 함수를 실행하라'''

if __name__ == "__main__":
    main()
    
